File: ios/PacketTunnel/XrayRunner.swift
import Foundation
import NetworkExtension

final class XrayRunner: EmbeddedPacketTunnelEngine {
  let name = "xray-core"
  private var sessionId: String?

  func start(packetFlow: NEPacketTunnelFlow, config: PacketTunnelConnectConfig) throws {
    let configJSON = try XrayConfigBuilder.build(from: config)
    sessionId = try EmbeddedCoreBridge.startXray(configJSON: configJSON, packetFlow: packetFlow)
  }

  func stop() {
    guard let sessionId else { return }
    EmbeddedCoreBridge.stopXray(sessionId: sessionId)
    self.sessionId = nil
  }
}

File: android/app/src/main/kotlin/com/example/inet_app/vpn/XrayRunner.kt
package com.example.inet_app.vpn

import android.net.VpnService

class XrayRunner : EmbeddedVpnEngine {
    override val name: String = "xray-core"
    private var sessionId: String? = null

    override fun start(service: VpnService, tunFd: Int, config: VpnConnectConfig) {
        val configJson = XrayConfigFactory.build(config)
        sessionId = EmbeddedCoreBridge.startXray(
            configJson = configJson,
            tunFd = tunFd,
            protectSocket = { socketFd -> service.protect(socketFd) },
        )
    }

    override fun stop() {
        val current = sessionId ?: return
        EmbeddedCoreBridge.stopXray(current)
        sessionId = null
    }
}
